src/treemodel/operation/attributeoperation.js

```diff
@@ -87,13 +88,22 @@ export default class AttributeOperation extends Operation {
 			 */
 			throw new CKEditorError(
 				'operation-attribute-different-keys: Old and new attributes should have the same keys.',
-				{ oldAttr: oldAttr, newAttr: newAttr } );
+				{ oldAttr: oldAttr, newAttr: newAttr }
+			);
 		}
 
+		// Split text nodes if needed, then get the nodes in the range and convert it to node list. It will be easier to operate.
+		this.range.start.parent._children._splitNodeAt( this.range.start.offset );
+		this.range.end.parent._children._splitNodeAt( this.range.end.offset );
+
 		// Remove or change.
 		if ( oldAttr !== null ) {
-			for ( let node of this.range.getAllNodes() ) {
-				if ( !node.hasAttr( oldAttr ) ) {
+			for ( let node of this.range.getAllNodes( true ) ) {
+				if ( node instanceof TextFragment ) {
+					node = node.first._nodeListText;
```

Comment:
This is probably the only place where such hack will be used so I don't know if it is a good idea to introduce a function for it. I understand that using `.first._nodeListText` is a bit confusing but if we introduce method in `NodeList` we would do even more complicated things: `node.first.parent._children._setAttrs( start, end, attrs )`. Keep in mind that we don't get pure node lists when iterating over a range. We could reorganize code / loop to extract node lists and then operate on them but it would probably be complicated too.

If anything, I'd introduce a method for `TextFragment` but I wanted to avoid calling `_nodeListText` there as that was supposed to be "almost private" entity for "`NodeList` use only".
`node.first.parent.